[NPM] bump gridplus-sdk to: v1.3.5

KevinVitale opened the pull request:
Fixes SDK issues related to upcoming Lattice1 firmware (v0.15.0).

### ToDo's
 - [x] Fix syntax to adhere to upstream's coding standards
 - [x] Make `baseUrl` _public_
 - [x] Include `0x` prefix on addresses
 - [x] Rebase with [beta.19](https://github.com/floating/frame/tree/beta.19)

douglance committed:
[NPM] 'gridplus-sdk' -> v1.3.2

Fixes SDK issues related to upcoming Lattice1 firmware (v0.15.0).

alex-miller-0 commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +265,25 @@ export default class Lattice extends Signer {
-      const signedTx = await sign(latticeTx, async tx => {
+      const signedTx: any = await sign(latticeTx, async tx => {
         const unsignedTx = this.createTransaction(index, rawTx.type, latticeTx.chainId, tx)
         const signOpts = { currency: 'ETH', data: unsignedTx }
```
Please use `getFwVersion` to inform how you will build this request.

Just because we worked to sanitize outputs from 0.15.0 bridged ETH requests doesn't mean we shouldn't put in the legwork here to avoid making bridged requests when fw is >=0.15.0. We don't want to maintain that shim forever.

alex-miller-0 replied:
Another thing to keep in mind is the ABI data. In 0.15.0 it will not be decoded unless you include it in the request. See this example from the docs: https://gridplus.github.io/gridplus-sdk/signing#1%EF%B8%8F%E2%83%A3--evm

We should basically replicate the transaction builder logic in `eth-lattice-keyring`: https://github.com/GridPlus/eth-lattice-keyring/blob/master/index.js#L192

KevinVitale committed:
[DEV] don't open dev tools

KevinVitale committed:
[NPM] run 'setup'

douglance committed:
Adds general signing

KevinVitale committed:
[DEV] restores default dev tool behavior

alex-miller-0 commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -317,13 +342,16 @@ export default class Lattice extends Signer {
     const signature = [
       sig.r,
       sig.s,
-      padToEven(sig.v.toString('hex'))
```
Are you sure this works? Seems like you should still need a `v` component...

KevinVitale replied:
This should be reverted.

alex-miller-0 commented on main/signers/Signer/index.ts:
```diff
@@ -77,7 +77,7 @@ export default class Signer extends EventEmitter {
       type: this.type,
       model: this.model,
-      addresses: this.addresses,
+      addresses: this.addresses.map(addr => addr.toString()),
```
This may not work as you think it does. If addresses are returned as buffers they will be missing the `0x` prefix, which I believe is required by Frame? As long as you are requesting ETH addresses and not pubkeys it shouldn't matter either way, but just pointing it out.

KevinVitale replied:
Frame displays addresses returned with a `0x` prefix, meaning it's being adding elsewhere. Seems that's sufficient and we don't need to add it ourselves.

alex-miller-0 replied:
Yeah I know, but I brought it up because here we are changing code that converts to strings. I think that code is unnecessary because that would only be required for pubkeys, which are returned as buffers. Addresses (which come with 0x prefix) are returned as strings, so you don't need to re-cast here.

But it's fine either way

mholtzman replied:
I think the `0x` prefix needs to be added in all places in this class otherwise there is potential for regression. while Frame adds the prefix for display purposes there are other instances in the code where these addresses may be used for things like string comparison that could break without the expected prefix.

alex-miller-0 replied:
The `0x` prefix is automatically added whenever you make a request to get a formatted address (which we have always done with Frame).

I brought it up because in a recent version we overloaded this method to allow request of pubkeys, which are returned as buffers. This would require a modification to the request, so not relevant here.

We're going to split that out in a future SDK version. Just wanted to make note of the distinction here and point to the `addr.toString()` casting being unnecessary.

KevinVitale replied:
Added `0x` prefix with this [commit](https://github.com/floating/frame/pull/845/commits/34be0f1d46b648d4b2c62449651cfd6cc7aa1eb4). Thank you, @douglance 🥇

alex-miller-0 commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -175,20 +169,20 @@ export default class Lattice extends Signer {
-        this.addresses = [...this.addresses, ...loadedAddresses]
+        const loadedAddresses = await connection.getAddresses(req)
+        this.addresses = [...this.addresses, ...loadedAddresses].map((addr) =>
+          addr.toString()
```
Same comment as above. This should be fine in practice but stringifying does not add the `0x` prefix. May be worth throwing in a util to map to string and add `0x` if there is none.

ccing @douglance

KevinVitale replied:
Same as above...Frame is including the `0x` when it's displayed to the user.

alex-miller-0 commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +262,52 @@ export default class Lattice extends Signer {
+      const connection = this.connection as Client
       const compatibility = signerCompatibility(rawTx, this.summary())
       const latticeTx = compatibility.compatible ? { ...rawTx } : londonToLegacy(rawTx)
+      const signerPath = this.getPath(index)
```
You could remove this line and use `unsignedTx.signerPath` as was being done before.

KevinVitale replied:
The line's removed. The original line has been restored.

alex-miller-0 reviewed (commented):
Mostly good. A few minor comments.

douglance committed:
Addresses feedback in PR #845

alex-miller-0 reviewed (dismissed):
Looks good 👍

KevinVitale commented:
Pinging @floating for a final review

mholtzman commented:
> Pinging @floating for a final review

taking a look at this today. in the meantime do you mind rebasing this on to https://github.com/floating/frame/tree/beta.19?

mholtzman commented on package.json:
```diff
@@ -251,7 +250,9 @@
-      "ws>utf-8-validate": false
+      "ws>utf-8-validate": false,
+      "gridplus-sdk>Buffer": true,
+      "gridplus-sdk": true
```
can you expand on which scripts need to run here?

KevinVitale replied:
Set this to `false`. [Commit](https://github.com/floating/frame/pull/845/commits/bbe30a3fc6f6be1671eea779df4a1aaac6bba632)

mholtzman commented on main/signers/lattice/adapter.ts:
```diff
@@ -70,7 +70,7 @@ export default class LatticeAdapter extends SignerAdapter {
           needsUpdate = true
         }
-
+        //@ts-expect-error -- baseUrl was made private in gridplus-sdk@v1.2.4
```
is there any other way to detect if the base URL has changed?

KevinVitale replied:
@mholtzman we're pushing an update that will make `baseUrl` _public_ again.

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/f3493308dfe58669392ea6078dabf9f32ff33384).

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -65,7 +61,7 @@ export default class Lattice extends Signer {
   tag = ''
 
-  constructor (deviceId: string, name: string, tag: string) {
+  constructor(deviceId: string, name: string, tag: string) {
```
removed a space here that's consistent with our formatting standards

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/ebc0fe38035651a7cd9ef17cff9b9b132f1ea21b)

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -175,20 +169,20 @@ export default class Lattice extends Signer {
+        const loadedAddresses = await connection.getAddresses(req)
+        this.addresses = [...this.addresses, ...loadedAddresses].map((addr) =>
+          addr.toString()
+        );
```
no semicolon

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/56f350c8dd1520883137961f99941a24e3c745ac)

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +262,51 @@ export default class Lattice extends Signer {
+      const connection = this.connection as Client
       const compatibility = signerCompatibility(rawTx, this.summary())
       const latticeTx = compatibility.compatible ? { ...rawTx } : londonToLegacy(rawTx)
+          const fwVersion = connection.getFwVersion()
```
tab

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/7aa368f4f829510eec0419a41074958e2049eccb)

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +262,51 @@ export default class Lattice extends Signer {
+            tx.getMessageToSign(false) :
+            rlp.encode(tx.getMessageToSign(false))
+
+          const to =  tx.to?.toString() ?? undefined
```
extra space

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/d4a36cda53fdc41042e1dc3f7a02dc0ad0b00bd5)

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +262,51 @@ export default class Lattice extends Signer {
+            encodingType: Constants.SIGNING.ENCODINGS.EVM,
+            signerPath: unsignedTx.signerPath,
+            decoder: callDataDecoder?.def
+          };
```
semicolon

KevinVitale replied:
[Commit](https://github.com/floating/frame/pull/845/commits/26aadd3a49656c383fd02e9a3b76cf5fb79eceed)

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -268,21 +262,51 @@ export default class Lattice extends Signer {
-        const result = await clientSign(signOpts)
-        const sig = result.sig as Signature
+        if (!sig) {
+          throw new Error('Signing failed')
```
is there any extra context that can be obtained here to display to the user or for debugging purposes?

alex-miller-0 replied:
Tbh it isn't possible to return an empty sig. `connection.sign` will throw an error if it encounters one. I think this is just an extra extra sanity check.

mholtzman replied:
if its truly not possible then prefer to remove this and it implies that it can happen and could be confusing down the road to people less familiar w. the code

mholtzman commented on main/signers/lattice/Lattice/index.ts:
```diff
@@ -317,8 +341,12 @@ export default class Lattice extends Signer {
+    const sig = result?.sig
+
+    if (!sig) {
+      throw new Error('Signing failed')
```
same comment as above re: error info

mholtzman reviewed (commented):
left a couple questions/comments. also, it seems like the only main functionality changes here are in the signing of messages and transactions. do you have any information on what scenarios this might change and what we could test for regression issues?

KevinVitale committed:
[PR#845] removed semicolon

PR #845 Conversation:
https://github.com/floating/frame/pull/845#discussion_r887173193

KevinVitale committed:
[PR#845] adjusted whitespacing

PR #845 Conversation:
https://github.com/floating/frame/pull/845#discussion_r887174875

KevinVitale committed:
[PR#845] removed extra space

PR #845 Conversation:
https://github.com/floating/frame/pull/845#discussion_r887175485

KevinVitale committed:
[PR#845] removed semicolon

PR #845 Conversation:
https://github.com/floating/frame/pull/845#discussion_r887177416

KevinVitale committed:
[PR#845] whitespace consistent with code standards

PR #845 Conversation:
https://github.com/floating/frame/pull/845/files/363dbddf568c2c8bad171c23b74f531fa791942a#r887171253

douglance committed:
Adds hex prefixes to address strings

douglance committed:
Removes gridplus-sdk allowScripts

douglance committed:
Fixes ts-expect-error declarations

KevinVitale commented:
> in the meantime do you mind rebasing this on to https://github.com/floating/frame/tree/beta.19?

The proposed branch has been rebased with _beta.19_

KevinVitale commented:
We'd like to know if the Frame team would like us to squash/cleanup commits. Please advise.

mholtzman commented:
> We'd like to know if the Frame team would like us to squash/cleanup commits. Please advise.

no that's OK, we can squash when we merge. you mind changing the base branch for this PR to beta.19?

mholtzman commented:
merged in squashed commit 130f3377ba0bd6499631f1a30fb0fa10a382ef2e
